File: celestia-orchestrator/run_tests.py
import os
import sys
import logging
from pathlib import Path
from datetime import datetime
from dotenv import load_dotenv
from hcloud import Client
from hcloud.images.domain import Image
from hcloud.server_types.domain import ServerType
from hcloud.locations.domain import Location
from hcloud.ssh_keys.domain import SSHKey

# --- Logger Setup ---
log_dir = Path(__file__).parent / "logs"
log_dir.mkdir(exist_ok=True)
timestamp = datetime.now().isoformat().replace(":", "-").split('.')[0]
log_file = log_dir / f"orchestrator-py-{timestamp}.log"

logging.basicConfig(
    level=logging.INFO,
    format="[%(levelname)s]: %(message)s",
    handlers=[
        # ✅ FIX: Specify UTF-8 encoding for file handler
        logging.FileHandler(log_file, encoding="utf-8"),
        logging.StreamHandler(sys.stdout),
    ],
)

# --- Configuration ---
# Load environment variables from .env file
dotenv_path = Path(__file__).parent / ".env"
load_dotenv(dotenv_path=dotenv_path)

HETZNER_TOKEN = os.getenv("HETZNER_API_TOKEN")
BACKEND_API_URL = os.getenv("BACKEND_API_URL")
BACKEND_API_KEY = os.getenv("BACKEND_API_KEY")
GIT_REPO_URL = os.getenv("GIT_REPO_URL")
HETZNER_SSH_KEY_ID = os.getenv("HETZNER_SSH_KEY_ID")

TEST_REGIONS = ["fsn1", "nbg1", "hel1", "ash"]
SERVER_TYPE = "cpx11"
SERVER_IMAGE = "ubuntu-22.04"

# --- Validation ---
def validate_config():
    """Ensure all required environment variables are set."""
    # Temporarily removed HETZNER_SSH_KEY_ID from validation for testing
    required = {
        "HETZNER_API_TOKEN": HETZNER_TOKEN,
        "BACKEND_API_URL": BACKEND_API_URL,
        "BACKEND_API_KEY": BACKEND_API_KEY,
        "GIT_REPO_URL": GIT_REPO_URL,
    }
    missing = [key for key, value in required.items() if not value]
    if missing:
        logging.error("Missing required environment variables:")
        for key in missing:
            logging.error(f"   - {key}")
        logging.error("\nPlease check your .env file.")
        sys.exit(1)

# --- Main Orchestration ---
def run_orchestration():
    """Main function to create servers in all test regions."""
    # ✅ FIX: Removed emoji from log messages
    logging.info("Starting orchestration (using Python)...")
    validate_config()

    try:
        client = Client(token=HETZNER_TOKEN)
        logging.info("Hetzner client initialized.")

        # ✅ FIX: Specify UTF-8 encoding when reading file
        cloud_init_path = Path(__file__).parent / "cloud-init.sh"
        cloud_init_template = cloud_init_path.read_text(encoding="utf-8")
        
        results = []

        for region in TEST_REGIONS:
            server_name = f"latency-tester-py-{region}-{int(datetime.now().timestamp())}"
            logging.info(f"Creating server in {region}... (Name: {server_name})")

            # Replace placeholders in the cloud-init script
            final_cloud_init = (
                cloud_init_template.replace("__HETZNER_API_TOKEN__", HETZNER_TOKEN)
                .replace("__GIT_REPO_URL__", GIT_REPO_URL)
                .replace("__BACKEND_API_URL__", BACKEND_API_URL)
                .replace("__BACKEND_API_KEY__", BACKEND_API_KEY)
                .replace("__REGION__", region)
            )

            try:
                # Create the server using the hcloud-python library
                # ✅ TEMP FIX: SSH Keys are commented out to prove they are the source of the error
                response = client.servers.create(
                    name=server_name,
                    server_type=ServerType(name=SERVER_TYPE),
                    image=Image(name=SERVER_IMAGE),
                    location=Location(name=region),
                    # ssh_keys=[SSHKey(id=int(HETZNER_SSH_KEY_ID))],
                    user_data=final_cloud_init,
                    labels={"project": "celestia-latency-monitor", "region": region},
                )
                server = response.server
                logging.info(f"   - Server creation initiated (ID: {server.id})")
                results.append({"region": region, "success": True, "serverId": server.id})

            except Exception as e:
                logging.error(f"   - Failed to create server in {region}: {e}")
                results.append({"region": region, "success": False, "error": str(e)})

        # Summary
        logging.info("========================================")
        logging.info("Orchestration Summary:")
        successful = sum(1 for r in results if r["success"])
        failed = len(results) - successful
        logging.info(f"Successful creations: {successful}/{len(TEST_REGIONS)}")
        logging.info(f"Failed creations: {failed}/{len(TEST_REGIONS)}")

        if failed > 0:
            logging.warning("Failed regions:")
            for r in results:
                if not r["success"]:
                    logging.warning(f"   - {r['region']}: {r['error']}")
        
        logging.info("\nOrchestration complete!")
        logging.info("See full details in the log file.")

    except Exception as e:
        logging.error(f"A fatal error occurred during orchestration: {e}")
        sys.exit(1)

if __name__ == "__main__":
    run_orchestration()

File: celestia-orchestrator/run-tests.js
require('dotenv').config();
const { spawn } = require('child_process');
const fs = require('fs');
const path = require('path');
const winston = require('winston');

// --- Logger Setup ---
const logsDir = path.join(__dirname, 'logs');
if (!fs.existsSync(logsDir)) {
  fs.mkdirSync(logsDir);
}
const timestamp = new Date().toISOString().replace(/:/g, '-');
const logFile = path.join(logsDir, `orchestrator-${timestamp}.log`);

const logger = winston.createLogger({
  level: 'info',
  format: winston.format.combine(
    winston.format.timestamp(),
    winston.format.printf(({ timestamp, level, message }) => `${timestamp} [${level.toUpperCase()}]: ${message}`)
  ),
  transports: [
    new winston.transports.File({ filename: logFile }),
    new winston.transports.Console({
      format: winston.format.combine(
        winston.format.colorize(),
        winston.format.printf(({ level, message }) => `${level}: ${message}`) // Use winston's level for colorization
      ),
    }),
  ],
});

// --- Configuration ---
const HETZNER_TOKEN = process.env.HETZNER_API_TOKEN;
const BACKEND_API_URL = process.env.BACKEND_API_URL;
const BACKEND_API_KEY = process.env.BACKEND_API_KEY;
const GIT_REPO_URL = process.env.GIT_REPO_URL;
const SSH_KEY_ID = parseInt(process.env.HETZNER_SSH_KEY_ID, 10);

const TEST_REGIONS = ['fsn1', 'nbg1', 'hel1', 'ash'];
const SERVER_TYPE = 'cpx11';
const SERVER_IMAGE = 'ubuntu-22.04';

// --- Validation ---
function validateConfig() {
  const required = { HETZNER_TOKEN, BACKEND_API_URL, BACKEND_API_KEY, GIT_REPO_URL, SSH_KEY_ID };
  const missing = Object.entries(required).filter(([, value]) => !value).map(([key]) => key);

  if (missing.length > 0) {
    logger.error('❌ Missing required environment variables:');
    missing.forEach(key => logger.error(`   - ${key}`));
    logger.error('\nPlease check your .env file.');
    process.exit(1);
  }
  if (isNaN(SSH_KEY_ID)) {
    logger.error('❌ HETZNER_SSH_KEY_ID must be a number.');
    process.exit(1);
  }
}

// --- Library-Free Server Creation with VERBOSE cURL ---
function createServer(serverData, region) {
  const jsonData = JSON.stringify(serverData);
  
  // ✅ FIX: Added -v (verbose) and -i (include headers) to curl arguments
  const args = [
    '-v', '-i',
    '-X', 'POST',
    '-H', `Authorization: Bearer ${HETZNER_TOKEN}`,
    '-H', 'Content-Type: application/json',
    '-d', jsonData,
    'https://api.hetzner.cloud/v1/servers'
  ];

  return new Promise((resolve, reject) => {
    const command = process.platform === 'win32' ? 'curl.exe' : 'curl';
    const curl = spawn(command, args);

    let output = '';
    curl.stdout.on('data', (data) => { output += data.toString(); });
    curl.stderr.on('data', (data) => { output += data.toString(); }); // Capture stderr too for verbose output

    curl.on('close', (code) => {
      // ✅ FIX: Log the FULL raw output for debugging
      logger.info(`\n--- Raw cURL Output for region [${region}] ---\n${output}\n------------------------------------------`);

      if (code !== 0) {
        reject(new Error(`cURL process exited with code ${code}.`));
        return;
      }
      
      // Extract JSON body from the raw output (it comes after the headers)
      const jsonBodyMatch = output.match(/\{.*\}/s);
      if (jsonBodyMatch) {
        try {
          const response = JSON.parse(jsonBodyMatch[0]);
          if (response.error) {
            reject(new Error(response.error.message));
          } else {
            resolve(response);
          }
        } catch (e) {
          reject(new Error("Failed to parse JSON from cURL output."));
        }
      } else {
         reject(new Error("No JSON body found in cURL output. The API might have returned an empty response."));
      }
    });

    curl.on('error', (err) => {
      reject(new Error(`Failed to start cURL process: ${err.message}`));
    });
  });
}

// --- Main Orchestration ---
async function runOrchestration() {
  logger.info('🚀 Starting orchestration (VERBOSE MODE)...');
  validateConfig();

  try {
    const cloudInitTemplate = fs.readFileSync(path.join(__dirname, 'cloud-init.sh'), 'utf8');
    const results = [];

    for (const region of TEST_REGIONS) {
      const serverName = `latency-tester-${region}-${Date.now()}`;
      logger.info(`📍 Creating server in ${region}... (Name: ${serverName})`);

      const finalCloudInitScript = cloudInitTemplate
        .replace(/__HETZNER_API_TOKEN__/g, HETZNER_TOKEN)
        .replace(/__GIT_REPO_URL__/g, GIT_REPO_URL)
        .replace(/__BACKEND_API_URL__/g, BACKEND_API_URL)
        .replace(/__BACKEND_API_KEY__/g, BACKEND_API_KEY)
        .replace(/__REGION__/g, region);

      const serverOptions = {
        name: serverName,
        server_type: SERVER_TYPE,
        image: SERVER_IMAGE,
        location: region,
        ssh_keys: [SSH_KEY_ID],
        user_data: finalCloudInitScript,
        start_after_create: true,
        labels: { project: 'celestia-latency-monitor', region: region },
      };

      try {
        const response = await createServer(serverOptions, region);
        logger.info(`   ✅ Server creation initiated (ID: ${response.server.id})`);
        results.push({ region, success: true, serverId: response.server.id });
      } catch (error) {
        logger.error(`   ❌ Failed to create server in ${region}: ${error.message}`);
        results.push({ region, success: false, error: error.message });
      }
    }

    // Summary
    logger.info('========================================');
    logger.info('📊 Orchestration Summary:');
    const successful = results.filter(r => r.success).length;
    const failed = results.filter(r => !r.success).length;
    logger.info(`✅ Successful creations: ${successful}/${TEST_REGIONS.length}`);
    logger.info(`❌ Failed creations: ${failed}/${TEST_REGIONS.length}`);
    if (failed > 0) {
      logger.warn('Failed regions:');
      results.filter(r => !r.success).forEach(r => logger.warn(`   - ${r.region}: ${r.error}`));
    }
    logger.info('\n🎉 Orchestration complete!');
    logger.info(`See full details in the log file: ${logFile}`);

  } catch (error) {
    logger.error('❌ A fatal error occurred during orchestration:', error.message);
    process.exit(1);
  }
}

runOrchestration();